File: src/py/is_palindrome/stdin_json_adapter.py
"""Stdin JSON adapter for ``is_palindrome_cli --impl py`` (library only; not fixtures.cli)."""

from __future__ import annotations

import json
import sys
from typing import Any

from is_palindrome.palindrome import (
    PalindromeException,
    is_palindrome,
    is_palindrome_from_utf8,
)


def main() -> None:
    raw = sys.stdin.read()
    try:
        req: dict[str, Any] = json.loads(raw.strip())
    except json.JSONDecodeError as e:
        print(e, file=sys.stderr)
        sys.exit(2)

    mode = req.get("mode", "")
    custom_raw = req.get("custom", [])
    custom: set[int] | None = None
    if custom_raw:
        custom = set(custom_raw)

    if mode == "hex":
        hex_s = req.get("hex")
        if hex_s is None:
            print("missing hex", file=sys.stderr)
            sys.exit(2)
        data = bytes(int(hex_s[i : i + 2], 16) for i in range(0, len(hex_s), 2))
        r = is_palindrome(data, custom)
        print("true" if r else "false")
        sys.exit(0 if r else 1)
    if mode == "string":
        text = req.get("text")
        if text is None:
            print("missing text", file=sys.stderr)
            sys.exit(2)
        try:
            r = is_palindrome_from_utf8(text, custom)
        except PalindromeException as e:
            print(e.error_code, file=sys.stderr)
            print("Input contains a scalar value > U+007F.", file=sys.stderr)
            sys.exit(2)
        print("true" if r else "false")
        sys.exit(0 if r else 1)

    print("unknown mode", file=sys.stderr)
    sys.exit(2)


if __name__ == "__main__":
    main()

File: CLI/src/bin/stdin_json_adapter.rs
//! Stdin JSON adapter binary (JSON on stdin) for subprocess checks against the Rust library.

use is_palindrome::{is_palindrome, is_palindrome_from_utf8};
use serde::Deserialize;
use std::collections::HashSet;
use std::io::{self, Read};

#[derive(Deserialize)]
struct Req {
    mode: String,
    hex: Option<String>,
    text: Option<String>,
    custom: Vec<u8>,
}

fn main() {
    let mut buf = String::new();
    io::stdin().read_to_string(&mut buf).expect("stdin");
    let req: Req = match serde_json::from_str(buf.trim()) {
        Ok(r) => r,
        Err(e) => {
            eprintln!("{e}");
            std::process::exit(2);
        }
    };
    let custom: Option<HashSet<u8>> = if req.custom.is_empty() {
        None
    } else {
        Some(req.custom.iter().copied().collect())
    };
    match req.mode.as_str() {
        "hex" => {
            let hex = match req.hex {
                Some(h) => h,
                None => {
                    eprintln!("missing hex");
                    std::process::exit(2);
                }
            };
            let mut v = Vec::new();
            for i in (0..hex.len()).step_by(2) {
                let b = u8::from_str_radix(&hex[i..i + 2], 16).expect("hex pair");
                v.push(b);
            }
            let r = is_palindrome(&v, custom.as_ref());
            print!("{}\n", if r { "true" } else { "false" });
            std::process::exit(if r { 0 } else { 1 });
        }
        "string" => {
            let text = match req.text {
                Some(t) => t,
                None => {
                    eprintln!("missing text");
                    std::process::exit(2);
                }
            };
            match is_palindrome_from_utf8(&text, custom.as_ref()) {
                Ok(r) => {
                    print!("{}\n", if r { "true" } else { "false" });
                    std::process::exit(if r { 0 } else { 1 });
                }
                Err(e) => {
                    eprintln!("{}", e.code);
                    eprintln!("Input contains a scalar value > U+007F.");
                    std::process::exit(2);
                }
            }
        }
        _ => {
            eprintln!("unknown mode");
            std::process::exit(2);
        }
    }
}
